arch: arm: move stack switching back to ASM code

wentongwu opened the pull request:
With -O0 optimizion, gcc compiler doesn't inline "static inline"
    marked function. So when function call return from function
    set_and_switch_to_psp which is to switch sp from MSP to PSP, the
    ending "mov sp, r7" instruction will overwrite the just updated
    sp value(PSP) with the beginning stack pointer(should be MSP)
    stored in r7 register, so the switch doesn't happen. And it causes
    unpredictable problems in the initialization process, the backward
    analysis for this problem can be found on Github issue #15794.
    
    Also increase stack buffer when code coverage enabled.

Fixes: #15794.

ioannisg reviewed (commented):
Although this seems to work, I wonder if we should just revert to using inline assembly for setting and switching to PSP.

psidhu commented:
> Although this seems to work, I wonder if we should just revert to using inline assembly for setting and switching to PSP.

@ioannisg Why did we switch away from asm here anyways? I'm assuming it was something to do with maintainability, but I'm not totally sure.

ioannisg commented:
> 
> 
> > Although this seems to work, I wonder if we should just revert to using inline assembly for setting and switching to PSP.
> 
> @ioannisg Why did we switch away from asm here anyways? I'm assuming it was something to do with maintainability, but I'm not totally sure.

Yeah, it was an effort to make the code more readable, but I guess we didn't review all the cases, unfortunately. and ended up with this kind of issues. I favor reverting the switch-and-set-psp() to assembly, but other reviewers are welcome to comment on the current way of fixing the bug.

psidhu commented:
> Yeah, it was an effort to make the code more readable, but I guess we didn't review all the cases, unfortunately. and ended up with this kind of issues. I favor reverting the switch-and-set-psp() to assembly, but other reviewers are welcome to comment on the current way of fixing the bug.

I agree with you to revert to this back to asm instead of fixing it this way. This is the second time I've seen a project attempt to do this, then decide to backoff so as not to rely on the compiler.

I think there's much merit in sticking with asm, but add in a block comment the roughly equivalent C code if anyone can't understand the asm, but would like to still understand what's happening.

andrewboie commented:
I think we should go back to assembly. Switching the stack pointer within C code is incredibly perilous, one bad assumption by the compiler about what is on the stack causes everything to come crashing down. You could consider x86_64's approach as a model however, look for _switch_top / _switch_bottom

ioannisg commented:
> 
> 
> I think we should go back to assembly. Switching the stack pointer within C code is incredibly perilous, one bad assumption by the compiler about what is on the stack causes everything to come crashing down. You could consider x86_64's approach as a model however, look for _switch_top / _switch_bottom

Will you be able to provide the patch @wentongwu ?

wentongwu commented:
> > I think we should go back to assembly. Switching the stack pointer within C code is incredibly perilous, one bad assumption by the compiler about what is on the stack causes everything to come crashing down. You could consider x86_64's approach as a model however, look for _switch_top / _switch_bottom
> 
> Will you be able to provide the patch @wentongwu ?

Yes, I'm preparing the patch. Thanks.

ioannisg commented:
> 
> 
> > > I think we should go back to assembly. Switching the stack pointer within C code is incredibly perilous, one bad assumption by the compiler about what is on the stack causes everything to come crashing down. You could consider x86_64's approach as a model however, look for _switch_top / _switch_bottom
> > 
> > 
> > Will you be able to provide the patch @wentongwu ?
> 
> Yes, I'm preparing the patch. Thanks.

Thank you!

andrewboie commented:
@ioannisg please have another look

ioannisg commented on arch/arm/core/cortex_m/reset.S:
```diff
@@ -61,12 +63,44 @@ SECTION_SUBSEC_FUNC(TEXT,_reset_section,__start)
     bl _WdogInit
 #endif
 
+#ifdef CONFIG_INIT_STACKS
```
This is OK with me, I think we can just keep it as is. Although, still the C code should work, cause I don't see any SP manipulation.

ioannisg commented on arch/arm/core/cortex_m/reset.S:
```diff
@@ -61,12 +63,44 @@ SECTION_SUBSEC_FUNC(TEXT,_reset_section,__start)
+#ifdef CONFIG_INIT_STACKS
+    ldr r0, =_interrupt_stack
+    ldr r1, =0xaa
+    ldr r2, =CONFIG_ISR_STACK_SIZE
```
One issue to consider, @andrewboie , un-related to the patch: we might be losing bytes, here, because _interrupt_stack is defined using K_THREAD_STACK_DEFINE, which adds, possibly a guard area. But this is not taken into account in CONFIG_ISR_STACK_SIZE - that's the original ISR stack configuration.

I propose to _not_ define the interrupt stack with K_THREAD_STACK_DEFINE, or add some smart options in the macro, to not include the guard area. Otherwise, we should use K_THREAD_STACK_SIZEOF(_interrupt_stack), but that we can do if we keep the function in C code.

andrewboie replied:
Didn't the old code do this?
I don't think we should refector this in this PR

ioannisg replied:
The old code did not do this, we always had this problem, unfortunately.
Sure, we should not fix this in the PR, I stressed that this is un-related, but, actually, we need to discuss _how_ to fix it (and with respect, also, to how and if we refactor the STACK_DEFINE macros). I'll post this to an existing/new issue.

ioannisg replied:
Resolving this - I am tracking down the issue with the ISR size in #16210

ioannisg commented on arch/arm/core/cortex_m/reset.S:
```diff
@@ -61,12 +63,44 @@ SECTION_SUBSEC_FUNC(TEXT,_reset_section,__start)
+     * gets set to _interrupt_stack during initialization.
+     */
+    ldr r0, =_interrupt_stack
+    ldr r1, =CONFIG_ISR_STACK_SIZE
```
Same issue here, with the beginning of the interrupt stack.

ioannisg replied:
Tracked in #16210

ioannisg commented on arch/arm/core/cortex_m/reset.S:
```diff
@@ -61,12 +63,44 @@ SECTION_SUBSEC_FUNC(TEXT,_reset_section,__start)
+    ldr r1, =CONFIG_ISR_STACK_SIZE
+    adds r0, r0, r1
+    msr PSP, r0
+    movs.n r0, #2
```
It's probably not a big deal - but you could do this as
Can you, instead, do this exactly as the C code suggested, 
`__set_CONTROL(__get_CONTROL() | CONTROL_SPSEL_Msk);`
i.e. you "or" the SPSEL bit instead of explicitly setting.

And, pls, add some inline comment, e.g. `/* __set_CONTROL(__get_CONTROL() | CONTROL_SPSEL_Msk); */` 
or simply /* CONTROL_SPSEL_Msk */

wentongwu replied:
yes, I will, Thanks

ioannisg reviewed (changes requested):
Did a pass - left some review comments, and a question for @andrewboie

wentongwu commented on arch/arm/core/cortex_m/reset.S:
```diff
@@ -61,12 +63,44 @@ SECTION_SUBSEC_FUNC(TEXT,_reset_section,__start)
 #endif
 
     /* lock interrupts: will get unlocked when switch to main task */
-    bl	lock_interrupts
```
for unify lock interrupt, how about pendsv, it does the same as patch does?

ioannisg replied:
I need to see what the pendSV does. In principle, I agree to unify this.
One thing about pendSV, though, is that it needs to be as fast as possible, while the boot code runs only once so we can allow some extra overhead to make the code look nice.

wentongwu replied:
yes, z_arch_irq_lock is higher overhead than direct ASM, so I mentioned pendsv. 
and z_arch_irq_lock is always inline.

ioannisg replied:
Let's keep the code as is - no big deal, actually. 

Ah, right, and, actually, it is defined only for` !ASMLANGUAGE`

ioannisg commented on arch/arm/core/cortex_m/reset.S:
```diff
@@ -61,12 +63,49 @@ SECTION_SUBSEC_FUNC(TEXT,_reset_section,__start)
+     * CONTROL_SPSEL_Msk
+     */
+    mrs r0, CONTROL
+    movs r1, #2
```
Pls, add the /* CONTROL_SPSEL_Msk */ comment in this line

wentongwu replied:
yes, it seems ARMV7-m can orrs immediate, but ARMV6-M can not (http://infocenter.arm.com/help/index.jsp?topic=/com.arm.doc.dui0662a/CIHBDBJH.html )

ioannisg reviewed (commented):
just a minor suggestion for the inline comment

wentongwu committed:
arch: arm: switching stack pointer with assembly code

With -O0 optimizion, gcc compiler doesn't inline "static inline"
marked function. So when function call return from function
set_and_switch_to_psp which is to switch sp from MSP to PSP, the
ending "mov sp, r7" instruction will overwrite the just updated
sp value(PSP) with the beginning stack pointer(should be MSP)
stored in r7 register, so the switch doesn't happen. And it causes
unpredictable problems in the initialization process, the backward
analysis for this problem can be found on Github issue #15794.

Fixes: #15794.

Signed-off-by: Wentong Wu <wentong.wu@intel.com>

wentongwu committed:
tests: kernel: increase stack buffer when code coverage enabled

increase stack buffer when code coverage enabled.

Fixes: #15794.

Signed-off-by: Wentong Wu <wentong.wu@intel.com>

manochavikas commented:
> With -O0 optimizion, gcc compiler doesn't inline "static inline" marked function

just saw this revert, an easy fix seems to be using always inline (__always_inline).

Cheers,
Vikas

manochavikas commented:
@wentongwu 
Can you check with always_inline, it should fix the issue with -O0.

Cheers,
Vikas
